rcl/test/rcl/test_get_type_description_service.cpp

```diff
@@ -51,87 +53,61 @@ static bool string_in_array(rcutils_string_array_t * array, const char * pattern
 
 static bool service_exists(
   const rcl_node_t * node_ptr, const char * service_name,
-  const char * service_type)
+  const char * service_type, std::chrono::milliseconds timeout)
 {
   rcl_allocator_t allocator = rcl_get_default_allocator();
 
-  rcl_names_and_types_t * srv_names_and_types =
-    static_cast<rcl_names_and_types_t *>(allocator.allocate(
-      sizeof(rcl_names_and_types_t),
-      allocator.state));
-  if (nullptr == srv_names_and_types) {
-    return false;
-  }
-  EXPECT_EQ(RCL_RET_OK, rcl_names_and_types_init(srv_names_and_types, 0, &allocator));
-  srv_names_and_types->names.data = NULL;
-  srv_names_and_types->names.size = 0;
-  srv_names_and_types->types = NULL;
-
-  OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
-  {
-    EXPECT_EQ(RCL_RET_OK, rcl_names_and_types_fini(srv_names_and_types));
-    allocator.deallocate(srv_names_and_types, allocator.state);
-  });
-
-  if (
-    RCL_RET_OK != rcl_get_service_names_and_types(
-      node_ptr,
-      &allocator, srv_names_and_types))
-  {
-    return false;
-  }
+  // Wait for a maximum of timeout seconds for the service to show up
+  auto start_time = std::chrono::system_clock::now();
+  while (std::chrono::system_clock::now() - start_time < timeout) {
+    rcl_names_and_types_t srv_names_and_types = rcl_get_zero_initialized_names_and_types();
 
-  if (srv_names_and_types->names.size < 1) {
-    return false;
-  }
-
-  const bool srv_name_found = string_in_array(
-    &srv_names_and_types->names,
-    service_name);
+    if (
+      RCL_RET_OK != rcl_get_service_names_and_types(
+        node_ptr,
+        &allocator, &srv_names_and_types))
+    {
+      return false;
```

Comment:
`OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT` should be called before this if that returns false?

Reply:
It isn't actually necessary in this case.

When you call `rcl_get_zero_initialized_names_and_types`, it returns a copy of the statically-zeroed structure, which we are storing on the stack.  When you call `rcl_get_service_names_and_types`, it expects to receive a zeroed `rcl_names_and_types_t` structure, and actually fails if it does not.  So if you call `rcl_names_and_types_init` first, `rcl_get_service_names_and_types` will fail.  So you only need to call `rcl_names_and_types_fini` after a successful `rcl_get_service_names_and_types`.

To be honest, I don't really like this API.  It operates differently from most of the rest of our APIs, which typically assume that you have run `init` first before doing anything else with the structure.  But it would be difficult to change (since it goes through the RMW layer), and is orthogonal to this change anyway.

Does this make sense to you?

Reply:
understood, thanks for the explanation.